### PR TITLE
Enode cleanup

### PR DESCRIPTION
**Pull request by aehyvari**
This PR suggests mostly minor cleanups in the egraph system.  I tried to organise the commits thematically so that they're easier to check.

### CODE REVIEW

File: src/tsolvers/egraph/EnodeStore.cc [hunk 1]
```diff
@@ -150,9 +150,7 @@ vec<PTRefERefPair> EnodeStore::constructTerm(PTRef tr) {
...
 
 bool EnodeStore::needsRecursiveDefinition(PTRef tr) const {
-    bool recdef = true;
-    for (auto ch : logic.getPterm(tr)) {
-        recdef &= needsEnode(ch);
-    }
-    return recdef;
+    Pterm const & t = logic.getPterm(tr);
+    opensmt::span const sp(t.begin(), t.size());
+    return std::all_of(sp.begin(), sp.end(), [&](PTRef ch) { return needsEnode(ch); });
```
Comment (blishko) [inline src/tsolvers/egraph/EnodeStore.cc:155]:
```suggestion
    return std::all_of(sp.begin(), sp.end(), [this](PTRef ch) { return needsEnode(ch); });
```
I like restricting the lambda's capture, if it is not too wordy. But I am also OK with your version.

File: src/tsolvers/egraph/EnodeStore.cc [hunk 2]
```diff
@@ -150,9 +150,7 @@ vec<PTRefERefPair> EnodeStore::constructTerm(PTRef tr) {
 }
 
 bool EnodeStore::needsRecursiveDefinition(PTRef tr) const {
-    bool recdef = true;
-    for (auto ch : logic.getPterm(tr)) {
-        recdef &= needsEnode(ch);
-    }
-    return recdef;
+    Pterm const & t = logic.getPterm(tr);
+    opensmt::span const sp(t.begin(), t.size());
```
Comment (blishko) [inline src/tsolvers/egraph/EnodeStore.cc:154]: Do we actually need the span here? We could use `t.begin()` and `t.end()` directly in `std::all_of`, no?

File: src/tsolvers/egraph/Explainer.cc [hunk 3]
```diff
@@ -48,10 +48,10 @@ void Explainer::storeExplanation(ERef x, ERef y, PtAsgn reason)
     // balanced (which is a requirement to keep the O(nlogn) bound
 
     // Make sure that x is the node with the larger number of edges to switch
-    const Enode& root_x = getEnode(getEnode(x).getRoot());
-    const Enode& root_y = getEnode(getEnode(y).getRoot());
+    Enode const & root_x = getEnode(getEnode(x).getRoot());
+    Enode const & root_y = getEnode(getEnode(y).getRoot());
     if ( root_x.getSize() < root_y.getSize() ) {
-        swap(x,y);
+        std::swap(x,y);
```
Comment (blishko) [inline src/tsolvers/egraph/Explainer.cc:54]: Seeing this now, I got a little worried. I thing we should not create the named references `root_x` and `root_y` and just inline them in the `if` check.
Because if the `swap` does occur, suddenly `root_x` is the root of `y` and `root_y` is the root of `x`, which is super confusing.
Hopefully we are not using these references later...

Comment (blishko) [reply src/tsolvers/egraph/Explainer.cc:54]: Also, this made me realize that I have inadvertently changed the meaning of this code in #314. Because `getSize()` now returns the number of children and this should now be `getEqSize()`.

**Review by blishko — COMMENTED** [review]
You changes uncovered an unwanted change I made in #314. We need to take a closer look, but I have to go now and will return to this later.

File: src/tsolvers/egraph/Enode.h [hunk 4]
```diff
@@ -33,6 +33,7 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #include "PtStructs.h"
 #include "SymRef.h"
 #include "TypeUtils.h"
+#include "Global.h"
```
Comment (blishko) [inline src/tsolvers/egraph/Enode.h:36]: Hopefully we do not need this include?

File: src/tsolvers/egraph/EnodeStore.cc (hunk 2 above)
Comment (aehyvari) [reply src/tsolvers/egraph/EnodeStore.cc:154]: Fixed.

File: src/tsolvers/egraph/Explainer.cc (hunk 3 above)
Comment (aehyvari) [reply src/tsolvers/egraph/Explainer.cc:54]: > Hopefully we are not using these references later...

They're not used afterwards.  I could make a scope just for this to make it clearer?

> Also, this made me realize that I have inadvertently changed the meaning of this code in #314. Because getSize() now returns the number of children and this should now be getEqSize().

I'll fix it in a follow-up commit.  If I understood you correctly, here it's enough to change `getSize()` => `getEqSize()`?

File: src/tsolvers/egraph/EnodeStore.cc (hunk 1 above)
Comment (aehyvari) [reply src/tsolvers/egraph/EnodeStore.cc:155]: For me it's the same.  We can go with your suggestion.  I think the three extra characters will not cause major problems with modern disk capacity.

**Review by blishko — COMMENTED** [review]
I think this looks good!

File: src/tsolvers/egraph/Enode.h (hunk 4 above)
Comment (blishko) [reply src/tsolvers/egraph/Enode.h:36]: This is needed because of the `SET_BIT` thing, I guess. Could we avoid this somehow?

File: src/tsolvers/egraph/Explainer.cc (hunk 3 above)
Comment (blishko) [reply src/tsolvers/egraph/Explainer.cc:54]: Yes, but I would like to measure the impact of this and also check that the other usages are correct.

Comment (blishko) [reply src/tsolvers/egraph/Explainer.cc:54]: > They're not used afterwards. I could make a scope just for this to make it clearer?

Yes, scope is also fine.

File: src/tsolvers/egraph/Enode.h (hunk 4 above)
Comment (aehyvari) [reply src/tsolvers/egraph/Enode.h:36]: Fixed.

**Review by blishko — APPROVED** [review]
I think this is all good now!
